### ISSUE TITLE
WebGL contexts not being cleaned up on destroy() [chrome browser]

### ISSUE THREAD

**Issue by AnonymousSausage77** [mistic100/Photo-Sphere-Viewer#1438]
### Describe the bug

Users in my application can select different tours to visit (ie. virtual tour plugin) which they can select from a menu. When they do this enough in chrome, it errors. 

When they choose a new tour, I am refreshing the whole plugin as follows:

I am using NextJS to control the lifecycle of my Photo Sphere Viewer instance. In a useEffect(), i am registering a new photo sphere viewer instance and then, on the cleanup of that useEffect, i am destroying the viewer by calling viewer.destroy(). Therefore, whenever the user closes and reopens the pano, this component remounts, the useEffect runs and the new viewer is recreated.

In my local dev environment, there is no issue in chrome; you can run this code no issue as many times in 'npm run dev' as you'd like. 
When built and deployed to my server, if you refresh more than 15 times, it crashes after ~15 times because the webgl contexts are not being destroyed despite my calling viewer.destroy(). I get:

fb7d5399.39fb0d39ba29a80f.js:230 WARNING: Too many active WebGL contexts. Oldest context will be lost.
9997.0a9e7fe6ff5d55aa.js:1 WebGL context lost
fb7d5399.39fb0d39ba29a80f.js:230 WARNING: Too many active WebGL contexts. Oldest context will be lost.
fb7d5399.39fb0d39ba29a80f.js:230 THREE.WebGLRenderer: Context Lost.
fb7d5399.39fb0d39ba29a80f.js:230 WARNING: Too many active WebGL contexts. Oldest context will be lost.
fb7d5399.39fb0d39ba29a80f.js:230 THREE.WebGLRenderer: Context Lost.

I know most browsers can only maintain 16 webgl context instances. I can only think of two things that this could mean:
- when deployed and accessed via chrome, PSV is not properly destroying the webgl context when it is reloaded (either because it is caching the context past refresh, or because viewer.destroy() is bugged)
- the local nextjs dev environment is handling webgl differently? idk, I'm scratching at possibilities.

### Photo Sphere Viewer version

5.8.3

### Plugins loaded

virtual-tour-plugin;equirectangular-tiles-adapter

### OS & browser

Apple M2 Pro; 13.4 (22F66)

### Additional context

Happens on **build** when on chrome (windows os), chrome (mac os).

**Does not** happen on firefox (mac os), nor on Safari (mac os)

To me, this seems like a chrome-related issue.

**Comment by AnonymousSausage77** [comment]
Ok, so, to attempt to fix this issue, i did something like:

```
// this does not rerender throughout my entire program

function Parent() {

    // while PSVWrapper may rerender, canvas does not which means
    // the webgl context should remain the same
    const canvas = useMemo(() => <canvas id="psv-canvas" />

    <PSVWrapper>
         {canvas}
    </PSVWrapper>
}
```

```

function PSVWrapper({children}) {

    useEffect(() => {
    
    // ....

      const canvasElement = document.getElementById("psv-canvas") as HTMLCanvasElement;
      const webGlContext = canvasElement.getContext("webgl2") || canvasElement.getContext("webgl");
  
      if (!webGlContext) {
        throw new Error("Failed to get webgl context")
      }
  
      const webglRendererParameters : WebGLRendererParameters = {
        canvas: canvasElement,
        context: webGlContext,
      }
  
      const newViewer = new PhotoSphereViewer({
        adapter: EquirectangularTilesAdapter,
        container: viewerContainerRef.current,
        zoomSpeed: 0.4,
        minFov: MIN_FOV,
        maxFov: MAX_FOV,
        rendererParameters: webglRendererParameters,
        
        navbar: false,
        plugins: [[VirtualTourPlugin, tourConfig]],
      });
      
      // ....
      
    }, [])
    
   return <div id="psv-container">
        {children}
    </div>
    
    }
```

**Comment by mistic100** [comment]
Instead of destroying an recreating the viewer shy don't use use the various methods to update the configuration ? Like `setNodes`.

Even if you use a service side datasource, you could simply change your datasource config then call `setCurrentNode` without rebuilding everything.

**Comment by mistic100** [comment]
I will add a `renderer.dispose()` call in the destroy. Though I am convinced the viewer should not be reloaded everytime (and probably not declared inside `useEffect` as far as I understand React lifecycle)

**Comment by AnonymousSausage77** [comment]
Cheers for the response and help.

I agree with you that my design is flawed, though it would take an extensive rebuild to change. It came about because, originally I was doing as you say, but I kept running into issues (that at this point I think I have blocked out of my conscious mind). The easiest solution, at the time, seemed to be to just re-initialise the entire viewer so I could be certain there was no hanging state from previous re-renders. It's debounced at least, so there are fewer reloads.

That said, even if I were to do this re-build to maintain a single canvas instead of recreating the PSV instance, the issue would still be present because:
- My pano is embedded in a modal whose contents is destroyed whenever:
    - the user closes the modal
    - another, non-PSV item is loaded into the modal like a PNG (This modal is used to display many things, including PSV)
- after enough times of changing between different elements in the modal, this error would still occur

The only way I can foreseeably prevent this issue without your fix would be to render the canvas somewhere completely different and create a portal which 'moves' the canvas into the modal when it is needed. This comes with its own set of headaches that I would hate to explain to my boss.

Or, I suppose, I would have to perform a major update to the modal is handles loading content, hiding instead of destroying content.

That is future-me's problem :).

Anyway, cheers for the help!

**Comment by AnonymousSausage77** [comment]
@mistic100 I was looking into this more because I still had the issues and decided to do what you suggested: to stop deleting the canvas but reuse it. That said, is there a way to force PSV to re-fetch a node when using EquirectangularPanos? Currently, I have changed the data of a node and want to refetch it without changing which node is currently being looked at, but neither:
`tour.setCurrentNode(nodeId)`
nor 
`tour.init();`
triggers a refetch of the existing node. For example, let's say i have changed the direction of the links (arrows) of currently-viewed pano, or perhaps I have changed the offset of the pano image: these were part of the reason that initially resulted to recreating the entire PSV instance; by recreating the entire viewer, I could trigger a refetch of the current node and all of its configurations.
